appinventor/blocklyeditor/src/workspace_svg.js

```diff
@@ -589,18 +589,54 @@ Blockly.WorkspaceSvg.prototype.customContextMenu = function(menuOptions) {
   /**
    * Function that returns a name to be used to sort blocks.
    * The general comparator is the block.category attribute.
-   * In the case of 'Components' the comparator is the instanceName of the component if it exists
-   * (it does not exist for generic components).
    * In the case of Procedures the comparator is the NAME(for definitions) or PROCNAME (for calls)
+   * In the case of 'Components' the comparator is the type of the component
    * @param {!Blockly.Block} block the block that will be compared in the sortByCategory function
    * @returns {string} text to be used in the comparison
    */
   function comparisonName(block){
-    if (block.category === 'Component' && block.instanceName)
-      return block.instanceName;
-    if (block.category === 'Procedures')
+    if (block.category == 'Variables'){
+      return block.category;
+    }
+    if (block.category === 'Procedures'){
       return (block.getFieldValue('NAME') || block.getFieldValue('PROCNAME'));
-    return block.category;
+    }
+    // Floating blocks that are not Component
+    return block["type"];
+  }
+
+  /**
+   * Function used to compare two strings with text and numbers
+   * @param {string} a first string to be compared
+   * @param {string} b second string to be compared
+   * @returns {number} returns 0 if the strings are equal, and -1 or 1 if they are not
+   */
+  function compareStrTextNum(strA,strB){
```

Comment:
App Inventor allows unicode names for components. For example, switch to a different language and pull in a button. This function only allows for names that are alphanumeric (also, no underscore). Rather than do it this way, I would propose using the regex `/(.*?)([0-9]+)$/` which will fail if there isn't a trailing number. If it succeeds, you parse it similar to what you do it below.

Reply:
It might also be worth experimenting with [`localeCompare`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) here, as it seems to be able to do collated comparisons as an option and should sort correctly for non-English languages.